### ISSUE TITLE
Docker network label does not works

### ISSUE THREAD

**Issue by rolurq** [containous/traefik#4876]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

I started traefik with docker directly, and some other containers with the traefik labels. I compose file I set up the traefik labels like this

```
# ...
ports:
  - 9000:9000
labels:
  - "traefik.docker.network=mycompose_default"
  - "traefik.frontend.rule=Host:s3.local"
  - "traefik.port=9000"
# ...
```

### What did you expect to see?

That the access to http://s3.local was performed correctly

### What did you see instead?

A gateway timeout accessing s3.local. I think the problem is that the `traefik.docker.network` label is not being used.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.2
Codename:     maroilles
Go version:   go1.11.1
Built:        2018-10-04_01:44:36PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
debug = true
logLevel = "debug"

[api]
  dashboard = true

[http.middlewares]
  # use gzip compression on responses when allowed
  [http.middlewares.test-compress.Compress]

[ping]
  entryPoint = "http"
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
time="2019-05-14T15:52:16Z" level=debug msg="'504 Gateway Timeout' caused by: dial tcp 172.18.0.3:9000: i/o timeout"

time="2019-05-14T15:52:16Z" level=debug msg="vulcand/oxy/forward/http: Round trip: http://172.18.0.3:9000, code: 504, Length: 15, duration: 30.00093049s"

time="2019-05-14T15:52:16Z" level=debug msg="vulcand/oxy/forward/http: completed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"http\",\"Opaque\":\"\",\"User\":null,\"Host\":\"172.18.0.3:9000\",\"Path\":\"\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Accept-Language\":[\"es-CU,es;q=0.9,en-US;q=0.8,en;q=0.7,es-419;q=0.6\"],\"Cache-Control\":[\"max-age=0\"],\"Connection\":[\"keep-alive\"],\"Cookie\":[\"G_ENABLED_IDPS=google\"],\"Upgrade-Insecure-Requests\":[\"1\"],\"User-Agent\":[\"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.131 Safari/537.36\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"s3.local\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.17.0.1:51360\",\"RequestURI\":\"/\",\"TLS\":null}"
```

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of communication regulation.

Thanks for your interest in Traefik!

We dedicate the issue tracker to bug reports and feature requests only. My advanced AI tells me this one is neither of them.

I encourage your to seek community support: join our [Slack workspace](https://slack.traefik.io/) and reach out to us on the [#support channel](https://traefik.slack.com/messages/C0CDT22PJ/).

Another option is to use [Stack Overflow](https://stackoverflow.com/questions/tagged/traefik) and tag your question `traefik`.

You can of course double check [Traefik's documentation](https://docs.traefik.io/) :sweat_smile:

**Comment by rolurq** [comment]
How is this not a bug??? Is a configuration entry that isn't working as is supposed to do

**Comment by ziggemannen** [comment]
Remove the ports: section. You are in effect bypassing traefik and presenting the application directly with your current composefile

**ports:
  - 9000:9000**
labels:
  - "traefik.docker.network=mycompose_default"
  - "traefik.frontend.rule=Host:s3.local"
  - "traefik.port=9000"
# ..

**Comment by rolurq** [comment]
@ziggemannen I am not presenting directly on the bridge, compose is creating a network for those containers (the one in  the traefik.docker.network label you know?) nevertheless I removed the ports section and still getting gateway timeout.

**Comment by ziggemannen** [comment]
Can you post the complete compose files for both your app and traefik?

Your app needs to use a netork that traefik is on, so if your composefile create a new network for your app, no wonder it does not work.

**Comment by rolurq** [comment]
@ziggemannen If traefik needs to manually be added to the network then what does the traefik.docker.network label does?

**Comment by ziggemannen** [comment]
Let me rephrase, yes it WILL work, but I suspect that the name of the network being created does not match. Post the full composefile, and I'll look at it. I just tested it here and it works.

**Comment by rolurq** [comment]
```
version: "2"

services:
  s3:
    image: minio/minio
    container_name: mys3
    volumes:
      - ./s3_data:/data
    labels:
      - "traefik.docker.network=s3test_default"
      - "traefik.frontend.rule=Host:mys3.local"
      - "traefik.port=9000"
    environment:
      MINIO_ACCESS_KEY: access
      MINIO_SECRET_KEY: secret
      MINIO_DOMAIN: mys3.local
    command: server /data
```
The network is not explicitly declared, this compose file is in a directory called `s3test` that's why docker-compose creates that network
traefik is not started with a compose file, I did it manually
```
$ docker run -d -p 80:80 --volume /var/run/docker.sock:/var/run/docker.sock --label "traefik.frontend.rule=Host:traefik.local" --label "traefik.port=8080" --volume /home/rolurq/traefik.toml:/etc/traefik/traefik.toml --name traefik traefik --api --docker
```

**Comment by ziggemannen** [comment]
Two things to check:

1. if you do a docker network ls, does the name of the network match: s3test_default?
2. Does the TOML file have the exposedbydefault = false or true?

if false, add the label traefik.enable=true

**Comment by rolurq** [comment]
This is the output of `docker network ls`
```
NETWORK ID          NAME                  DRIVER              SCOPE
358653ee9993        s3test_default        bridge              local
0e87c5b9782c        bridge                bridge              local
9a1589f9ff5b        dev_flymariposa       bridge              local
521ce9560e38        host                  host                local
3a61aa620a32        none                  null                local
```
and the toml file is what is wrote in this issue, it has that content exactly

**Comment by ziggemannen** [comment]
Looking at your docker run for traefik, you have put front end rules on traefik itself? I wonder what that could cause.

**Comment by rolurq** [comment]
@ziggemannen to access the traefik daskboard

**Comment by dduportal** [comment]
Hi @rolurq . This is not a Traefik issue, but more a Docker-related due to networks:

By running Traefik with the `docker run` command you provided, then the Traefik container is bound to the default bridge network: you can verify this with the command `docker network inspect bridge`: you'll see an entry (in the pile of JSON) related to the Traefik container.

But as the `docker-compose` documentation states (https://docs.docker.com/compose/networking/):

> By default Compose sets up a single network for your app

It means that the file `s3test/docker-compose.yml` creates a dedicated network by default, for your application stack, as you can see with the result of your `docker network ls` command (there is a network named `s3test_default`).

So the `Gateway Timeout` error is expected: Traefik being in the network `bridge`, then it cannot reach the network `s3test_default`.

Different solutions to solve your issue:

- If you need to keep Traefik running with the `docker run` command you have, then you must switch  the network mode to `bridge` in your `s3test/docker-compose.yml` file (ref. https://docs.docker.com/compose/compose-file/compose-file-v2/).
It should looks like this:

```yaml
## Traefik launched with the following command:
# docker run -d -p 80:80 --volume /var/run/docker.sock:/var/run/docker.sock --label "traefik.frontend.rule=Host:traefik.localhost" --label "traefik.port=8080" --name traefik traefik --api --docker
version: "2"

services:
  s3:
    image: minio/minio
    container_name: mys3
    network_mode: bridge
    volumes:
      - ./s3_data:/data
    labels:
      - "traefik.frontend.rule=Host:mys3.localhost"
      - "traefik.port=9000"
    environment:
      MINIO_ACCESS_KEY: access
      MINIO_SECRET_KEY: secret
      MINIO_DOMAIN: mys3.local
    command: server /data
```

- Alternatively, you can define a dedicated network for Traefik and then reference it as a named network, as @ziggemannen showed earlier, in your application's `docker-compose.yml` file.

Let us know the result of your changes.

**Comment by rolurq** [comment]
@dduportal I get your point, and again doesn't the `traefik.docker.network` setting label exists for traefik to connect to the specified network?
I mean, the docker api provides a way to connect a container to a network so this can be done, taking into account that traefik uses the docker api. I though is what that label does, if that isn't the case, I think it should be cleared in the documentation.

**Comment by dduportal** [comment]
@rolurq you could also use the command `https://docs.docker.com/engine/reference/commandline/network_connect/` .
This is written in the documentation: access https://docs.traefik.io/configuration/backends/docker/#usebindportip and scroll up to check the "Warning" admonition.

The definition of the label `traefik.docker.network` states the following (from https://docs.traefik.io/configuration/backends/docker/#on-containers):
> If a container is linked to several networks, be sure to set the proper network name (you can check with docker inspect <container_id>) otherwise it will randomly pick one (depending on how docker is returning them).
For instance when deploying docker stack from compose files, the compose defined networks will be prefixed with the stack name. Or if your service references external network use it's name instead.

Keep in mind that Traefik is acting at layer 7 of the network (eventually layer 4 - transport with the upcoming `v2.0`), while the connectivity issue is happening at layer 3 (Routing / IP). Enabling Traefik to do such a change would be risky as it would Traefik to do whatever it wants on the network. The proposed solution with `docker network connect` (or the change I pointed you to) are solving first the routing issue at layer 3, with an administrative action. Then Traefik (and the label) is able to select the right route to contact the backend.

However, you're right by saying this could be improved a bit because for better visibility in the documentation. As you are in the status of "searching documentation and not finding this information", would you be willing to help us by contributing such an improvement?

**Comment by rolurq** [comment]
I used the command you mentioned, is what I did from the start, I just thought that traefik will do it automatically per the configuration mentioned.

I know that to make that automatically is risky, thats what I thought when I begin to think in the implementation issues.

Of course, I will be willing to contribute with improving the documentation.  I guess the quicker change could be rename the setting to `traefik.docker.preferred_network`, which has a little more clear that what it does is to select a network, not binding to it, but this will be a non backward compatible change.

**Comment by dduportal** [comment]
@rolurq renaming the settings is an heavy change, particularly in the context of the upcoming Traefik v2.0.

What do you think about improving the explanation of how "Traefik + Docker" works in regard to the network?

Also, the naming of the option does not conveys the "preferred" network, but the target network to be able to choose the correct route to reach the backend.
